design/config.txt
.. mode: -*- rst -*-

MPS Configuration
=================

:Tag: design.mps.config
:Author: Richard Brooksby
:Date: 1997-02-19
:Status: incomplete design
:Revision: $Id: //info.ravenbrook.com/project/mps/master/design/config.txt#15 $
:Copyright: See `Copyright and License`_.
:Index terms: pair: configuration; design


Introduction
------------

_`.intro`: This document describes how the `Memory Pool System
<http://www.ravenbrook.com/project/mps/>`_ source code is configured so
that it can target different architectures, operating systems, build
environments, varieties, and products.



Requirements
------------

_`.req.import`: The MPS must be simple to include in third-party projects.

_`.req.arch`: Allow architecture specific configurations of the MPS, so
that we can vary the MPS according to the target architecture.

_`.req.os`: Allow operating system specific configurations of the MPS,
so that we can vary the MPS according to the target OS.

_`.req.builder`: Allow build environment specific configurations of the
MPS, so that we can vary the MPS according to the compiler, etc.

_`.req.var`: Allow configurations with different amounts of
instrumentation (assertions, metering, etc.).

_`.req.impact`: The configuration system should have a minimal effect on
maintainability of the implementation.

_`.req.port`: The system should be easy to port across platforms.

_`.req.maint`: Maintenance of the configuration and build system should
not consume much developer time.


Retired requirements
....................

_`.req.prod`: Allow product specific configurations of the MPS, so that
we can build variants of the MPS for use in different products.  This
requirement has been retired on 2012-09-03 as part of work on the
variety-reform_ branch.  Client-specific customisation of the MPS will
be handled in source control, while the MPS source remains generic, to
reduce costs and increase reliability.  See [RB_2012-09-13]_.

.. _variety-reform: /project/mps/branch/2012-08-15/variety-reform


Definitions
-----------

_`.def.platform`: A *platform* is a combination of an architecture
(`.def.arch`_), an operating system (`.def.os`_), and a builder
(`.def.builder`_).  The set of supported platforms is maintained in the
`Platforms section of "Building the Memory Pool System"
<../manual/html/guide/build.html#platforms>`_.

_`.def.arch`: An *architecture* is processor type with associated calling
conventions and other binary interface stuff these days often called the
`ABI <http://en.wikipedia.org/wiki/Application_binary_interface>`_. 
Most importantly for the MPS it determines the layout of the register
file, thread context, and thread stack.

_`.def.os`: An *operating system* is the interface to external resources. 
Most importantly for the MPS it determines the low level interface to
virtual memory (if any) and threading.

_`.def.builder`: A *builder* is the tools (C compiler, etc.) used to make
the target (`.def.target`_).  The MPS minimises use of compiler-specific
extensions, but this is handy for suppressing warnings, inlining hints,
etc.

_`.def.var`: A *variety* determines things like the amount of debugging,
internal consistency checking, annotation, etc.  In modern IDEs this
called a "build configuration" and the usual default is to have two:
"debug" and "release". The MPS predates this convention, but the concept
is the same.

_`.def.prod`: A *product* is the intended product into which the MPS will
fit, e.g. ScriptWorks, Dylan, etc.  We no longer maintain this concept
as a dimension of configuration since `.req.prod`_ has been retired.

_`.def.target`: The *target* is the result of the build.


Overview
--------

_`.import.source`: The MPS can be simply included in client products as
source code.  Since `version 1.110`_ we made it possible to simply
include the file ``mps.c`` in a client's build process, without
requiring a separate build of the MPS or linking a library.  This is
described `section 2.3.1, "Compiling for production" of the MPS manual
<../manual/html/guide/build.html#compiling-for-production>`_.

.. _`version 1.110`: http://www.ravenbrook.com/project/mps/version/1.110/

_`.no-gen`: No generated code or external tools are required.  On most
platforms the only tool is the C compiler.  On 64-bit Windows we require
the assembler since Microsoft withdrew in-line assembler from their C
compiler.

_`.no-spaghetti`: Several of the MPS team have worked on some extremely
messy code bases which used a great number of ``#ifdef`` statements. 
These quickly became very expensive to maintain and develop.  The
general rule in the MPS is "no ``#ifdefs``".  Instead, platform-specific
code is kept in separate source files and selected by carefully controlled
``#ifdefs``, such as in `mps.c <../code/mps.c>`_.

_`.min-dep`: Dependency on a particular configuration should be
minimized and localized when developing code.  This is enshrined in the
general rules for implementation [ref?] that are enforced by MPS
development procedures including code review and inspection.


The build system
----------------

Abstract build function
.......................

_`.build.fun`: The MPS implementation assumes only a simple "build
function" that takes a set of sources, possibly in several languages,
compiles them with a set of predefined preprocessor symbols, and links
the result with a set of libraries to form the target::

    target := build(<defs>, <srcs>, <libs>)

_`.build.sep`: Separate compilation and linkage can be seen as a
memoization of this function, and is not strictly necessary for the
build. Indeed, since `version 1.110` we found that modern compilers
are quite happy to compile the whole MPS in one go `.import.source`_.

_`.build.cc`: A consequence of this approach is that it should always
be possible to build a complete target with a single UNIX command line
calling the compiler driver (usually "cc" or "gcc"), for example::

    cc -o main -DCONFIG_VAR_DF foo.c bar.c baz.s -lz

_`.build.defs`: The "defs" are the set of preprocessor macros which are to be 
predefined when compiling the module sources::

    CONFIG_VAR_<variety-code>

The variety-codes are the letter code that appears after "variety." in
the tag of the relevant variety document (see variety.*) converted to
upper case. Currently (2012-09-03):

_`.var.hot`: ``HOT``

    Intended for release in products.  Optimised, reduced internal
    checking, especially on the critical path [RB_2012-09-07]_.

_`.var.cool`: ``COOL``

    Intended for use during development.  Moderately thorough internal
    consistency checking.  Reduced optimisation to allow for
    single-stepping.

_`.var.rash`: ``RASH``

    No internal checking at all.  Slight performance improvement over
    `.var.hot`_ at the cost of early detection of memory management
    bugs.  We do not advise use of this variety, as memory management
    bugs tend to be extremely expensive to deal with.

_`.default.hot`: If no ``CONFIG_VAR`` is present, ``HOT`` is assumed in
`config.h`_.

_`.build.srcs`: The "srcs" are the set of sources that must be
compiled in order to build the target. The set of sources may vary
depending on the configuration. For example, different sets of sources
may be required to build different architectures.

.. note::

    This is a dependency between the makefile (or whatever) and the
    module configuration in `config.h`_.

_`.build.libs`: The "libs" are the set of libraries to which the
compiled sources must be linked in order to build the target. For
example, when building a test program, it might include the ANSI C
library and an operating system interface library.


File Structure
..............

_`.file.dir`: The MPS source code is arranged in a single directory
called "code" containing all the sources for the whole family of
targets.  

_`.file.base`: The names of sources must be unique in the first eight
characters in order to conform to FAT filesystem naming restrictions.
(Do not scoff -- this has been an important requirement as recently as
2012!)

_`.file.ext`: The extension may be up to three characters and directly
indicates the source language.

_`.file.platform`: Platform-specific files include the platform code
in their name.  See `.mod.impls`_.


Modules and naming
..................

_`.mod.unique`: Each module has an identifier which is unique within the MPS.  

_`.mod.impls`: Each module has one or more implementations which may be
in any language supported by the relevant build environment. 

_`.mod.primary`: The primary implementation of a module is written in
target-independent ANSI C in a source file with the same name as the
module.

_`.mod.an`: Where there are platform-specific implementations and an
inferior portable ANSI C fallback implementation, "an" is used in
place of the platform code.

_`.mod.secondary`: The names of other implementations should begin
with the same prefix (the module id or a shortened version of it) and
be suffixed with on or more target parameter codes (defined below). In
particular, the names of assembly language sources must include the
target parameter code for the relevant architecture.

_`.mod.example`: For example, the stack scanner is defined in `ss.h
<../code/ss.h>`_ (which is platform-independent).  It has some
platform-independent C in `ss.c <../code/ss.c>`_ and, for example,
`ssw3i6mv.c <../code/ssw3i6mv.c>`_ is specific to Windows on the x64
architecture built with Microsoft Visual C.


Build system rationale
......................

_`.build.rat`: This simple design makes it possible to build the MPS
using many different tools.  Microsoft Visual C and other graphical
development tools do not support much in the way of generated sources,
staged building, or other such stuff.  The Visual C and Xcode "project"
files correspond closely to a closure of the build function
(`.build.fun`_).  The simplicity of the build function has also made it
easy to set up builds using NMAKE (DOS), MPW (Macintosh), and to get the
MPS up and running on other platforms such as FreeBSD and Linux in very
little time.  The cost of maintaining the build systems on these various
platforms is also reduced to a minimum, allowing the MPS developers to
concentrate on primary development.  The source code is kept simple and
straightforward.  When looking at MPS sources you can tell exactly what
is going to be generated with very little context.  The sources are not
munged beyond the standard ANSI C preprocessor.

_`.build.port`: The portability requirement (`.req.port`_) implies that
the build system must use only standard tools that will be available on
all conceivable target platforms.  Experience of development
environments on the Macintosh (Metrowerks Codewarrior) and Windows NT
(Visual C++) indicates that we cannot assume much sophistication in the
use of file structure by development environments.  The best that we can
hope for is the ability to combine a fixed list of source files,
libraries, and predefined preprocessor symbols into a single target.

_`.build.maint`: The maintainability requirement (`.req.maint`_) implies
that we don't spend time trying to develop a set of tools to support
anything more complicated than the simple build function described
above.  The effort in constructing and maintaining a portable system of
this kind is considerable. Such efforts failed in the Electronic
Publishing division of Harlequin.


Implementation
--------------

_`.impl`: The two implementation files `config.h`_ and `mpstd.h`_ can be
seen as preprocessor programs which "accept" build parameters and "emit"
configuration parameters (`.fig.impl`_).  The build parameters are
defined either by the builder (in the case of target detection) or by
the build function (in the case of selecting the variety).

_`.fig.impl`:

=========================== ============== ===========================================
Build parameters            Source file    Configuration parameters
=========================== ============== ===========================================
``CONFIG_VAR_HOT``          ⟶ ``config.h`` ⟶ ``MPS_ASSERT_STRING``, etc.
``_WIN32``                  ⟶ ``mpstd.h``  ⟶ ``MPS_OS_W3``, etc.
=========================== ============== ===========================================

_`.impl.dep`: No source code, other than the directives in `config.h`_
and `mpstd.h`_, should depend on any build parameters.  That is,
identifers beginning "CONFIG\_" should only appear in impl.h.config. 
Code may depend on configuration parameters in certain, limited ways, as
defined below (`.conf`_).

.. _`config.h`: <../code/config.h>
.. _`mpstd.h`: <../code/mpstd.h>


Target platform detection
.........................

_`.pf`: The target platform is "detected" by the preprocessor directives in 
`mpstd.h`_.

_`.pf.form`: This file consists of sets of directives of the form::

    #elif <conjunction of builder predefinitions>
    #define MPS_PF_<platform code>
    #define MPS_OS_<operating system code>
    #define MPS_ARCH_<architecture code>
    #define MPS_BUILD_<builder code>
    #define MPS_T_WORD     <word type>
    #define MPS_T_ULONGEST <longest unsigned integer type>
    #define MPS_WORD_SHIFT <word shift>
    #define MPS_PF_ALIGN   <minimum alignment>

_`.pf.detect`: The conjunction of builder predefinitions is a constant
expression which detects the target platform.  It is a logical AND of
expressions which look for preprocessor symbols defined by the build
environment to indicate the target.  These must be accompanied by a
reference to the build tool documentation from which the symbols came. 
For example::

    /* "Predefined Macros" from "Visual Studio 2010" on MSDN
     * <http://msdn.microsoft.com/en-us/library/b0084kay(v=vs.100).aspx>. */

    #elif defined(_MSC_VER) && defined(_WIN32) && defined(_M_IX86)

_`.pf.codes`: The declarations of the platform, operating system,
architecture, and builder codes define preprocessor macros corresponding
the the target detected (`.pf.detect`_).  For example::

    #define MPS_PF_W3I3MV
    #define MPS_OS_W3
    #define MPS_ARCH_I3
    #define MPS_BUILD_MV

_`.pf.word`: The declaration of ``MPS_T_WORD`` defines the unsigned
integral type which corresponds, on the detected target, to the
machine word. It is used to defined the MPS Word type
(design.mps.type.word). For example::

    #define MPS_T_WORD      unsigned long

We avoid using ``typedef`` here because `mpstd.h`_ could potentially
be included in assembly language source code.

_`.pf.word-width`: The declaration of ``MPS_WORD_WIDTH`` defines the
number of bits in the type defined by ``MPS_T_WORD`` (`.pf.word`_) on the
target. For example::

    #define MPS_WORD_WIDTH  32

_`.pf.word-shift`: The declaration of ``MPS_WORD_SHIFT`` defines the log
to the base 2 of ``MPS_WORD_WIDTH``.  For example::

    #define MPS_WORD_SHIFT  5

_`.pf.pf-align`: The declaration of ``MPS_PF_ALIGN`` defines the minimum
alignment which must be used for a memory block to permit any normal
processor memory access.  In other words, it is the maximum alignment
required by the processor for normal memory access.  For example::

    #define MPS_PF_ALIGN    4

_`.pf.ulongest`: The declaration of ``MPS_T_ULONGEST`` defines the
longest available unsigned integer type on the platform.  This is
usually just ``unsigned long`` but under Microsoft C on 64-bit Windows
``unsigned long`` is just 32-bits (curse them!)  For example::

    #define MPS_T_ULONGEST      unsigned __int64

_`.pf.pf-string`: The declaration of ``MPS_PF_STRING`` defines a string
that is used to identify the target platform in `version.c
<../code/version.c>`_.  For example::

    #define MPS_PF_STRING   "w3i6mv"


Target varieties
................

_`.var`: The target variety is handled by preprocessor directives in
impl.h.config.

_`.var.form`: The file contains sets of directives of the form::

    #if defined(CONFIG_VAR_COOL)
    #define CONFIG_ASSERT
    #define CONFIG_ASSERT_ALL
    #define CONFIG_STATS

_`.var.detect`: The configured variety is one of the variety
preprocessor definitions passed to the build function
(`.build.defs`_), for example, ``CONFIG_VAR_COOL``. These are
decoupled in order to keep the number of supported varieties small,
controlling each feature (for example, assertions) by a single
preprocessor definition, and maintaining flexibility about which
features are enabled in each variety.

_`.var.symbols`: The directives should define whatever symbols are
necessary to control featrues. These symbols parameterize other parts
of the code, such as the declaration of assertions, etc. The symbols
should all begin with the prefix ``CONFIG_``.


Source code configuration
-------------------------

_`.conf`: This section describes how the configuration may affect the
source code of the MPS.

_`.conf.limit`: The form of dependency allowed is carefully limited to
ensure that code remains maintainable and portable (`.req.impact`_).

_`.conf.min`: The dependency of code on configuration parameters should
be kept to a minimum in order to keep the system maintainable
(`.req.impact`_).


Configuration Parameters
........................

_`.conf.params`: The compilation of a module is parameterized by::

    MPS_ARCH_<arch-code>
    MPS_OS_<os-code>
    MPS_BUILD_<builder-code>
    MPS_PF_<platform-code>


Abstract and Concrete Module Interfaces
.......................................

_`.abs.caller`: Basic principle: the caller musn't be affected by
configuration of a module. This reduces complexity and dependency of
configuration.  All callers use the same abstract interface.  Caller
code does not change.

_`.abs.interface`: Abstract interface includes:

- method definitions (logical function prototypes which may be macro methods)
- names of types
- names of constants
- names of structures and fields which form part of the interface, and 
  possibly their types, depending on the protocol defined
- the protocols

_`.abs.rule`: The abstract interface to a module may not be altered by a

code/config.h
/* config.h: MPS CONFIGURATION
 *
 * $Id: //info.ravenbrook.com/project/mps/master/code/config.h#52 $
 * Copyright (c) 2001-2013 Ravenbrook Limited.  See end of file for license.
 * Portions copyright (c) 2002 Global Graphics Software.
 *
 * PURPOSE
 *
 * This module translates from high-level symbols defined by the
 * external build system (gnumake, nmake, etc.) into specific sets
 * of features used by MPS modules.
 *
 * DESIGN
 *
 * See <design/config/>.
 */

#ifndef config_h
#define config_h


/* CONFIG_VAR_* -- variety Configuration
 *
 * These are translated into the directives CONFIG_ASSERT, CONFIG_STATS,
 * CONFIG_LOG, etc. which control actual compilation features.
 */

/* CONFIG_VAR_RASH -- the rash and reckless variety
 *
 * This variety switches off as many features as possible for maximum
 * performance, but is therefore unsafe and undebuggable.  It is not intended
 * for use, but for comparison with the hot variety, to check that assertion,
 * logging, etc. have negligible overhead.
 */

#if defined(CONFIG_VAR_RASH)
/* no asserts */
/* no statistic meters */
/* no telemetry log events */


/* CONFIG_VAR_COOL -- cool variety
 *
 * The cool variety is intended for use when developing an integration with
 * the MPS or debugging memory problems or collecting detailed telemetry
 * data for performance analysis.  It has more thorough consistency checking
 * and data collection and output, and full debugging information.
 */

#elif defined(CONFIG_VAR_COOL)
#define CONFIG_ASSERT
#define CONFIG_ASSERT_ALL
#define CONFIG_ASSERT_ABORT
#define CONFIG_STATS
#ifndef CHECKLEVEL
#define CHECKLEVEL      CheckLevelSHALLOW
#endif
#define CONFIG_LOG
#define CONFIG_LOG_ALL


#else /* CONFIG_VAR_* */

/* CONFIG_VAR_HOT -- the hot variety
 *
 * This variety is the default variety for distribution in products that use
 * the MPS.  It has maximum performance while retaining a good level of
 * consistency checking and allowing some debugging and telemetry features.
 */

/* #elif defined(CONFIG_VAR_HOT) */
#define CONFIG_ASSERT
/* Note, not CONFIG_ASSERT_ABORT */
#ifndef CHECKLEVEL
#define CHECKLEVEL      CheckLevelMINIMAL
#endif
/* no statistic meters */
#define CONFIG_LOG

#endif /* CONFIG_VAR_* */


/* Build Features */


#if defined(CONFIG_ASSERT)
/* asserts: AVER, AVERT, NOTREACHED, CHECKx */
/* note: a direct call to ASSERT() will *still* fire */
#define AVER_AND_CHECK
#if defined(CONFIG_ASSERT_ALL)
#define AVER_AND_CHECK_ALL
#define MPS_ASSERT_STRING "assertastic"
#else /* CONFIG_ASSERT_ALL, not */
#define MPS_ASSERT_STRING "asserted"
#endif /* CONFIG_ASSERT_ALL */
#else /* CONFIG_ASSERT, not */
#define AVER_AND_CHECK_NONE
#define MPS_ASSERT_STRING "nonasserted"
#endif
#if defined(CONFIG_ASSERT_ABORT)
#define ASSERT_ABORT() abort()
#else
#define ASSERT_ABORT() NOOP
#endif


#if defined(CONFIG_STATS)
/* CONFIG_STATS = STATISTICS = METERs */
/* WARNING: this may change the size and fields of MPS structs */
/* (...but see STATISTIC_DECL, which is invariant) */
#define STATISTICS
#define MPS_STATS_STRING "stats"
#else
#define STATISTICS_NONE
#define MPS_STATS_STRING "nonstats"
#endif


#if defined(CONFIG_LOG)
/* TELEMETRY = LOG = EVENTs */
#define EVENT
#if defined(CONFIG_LOG_ALL)
#define EVENT_ALL 1     /* log events on critical path */
#define MPS_LOG_STRING "logtastic"
#else /* CONFIG_LOG_ALL, not */
#define EVENT_ALL 0     /* don't log events on critical path */
#define MPS_LOG_STRING "logging"
#endif /* CONFIG_LOG_ALL */
#else /* CONFIG_LOG, not */
#define EVENT_NONE
#define MPS_LOG_STRING "nonlogging"
#endif /* CONFIG_LOG */


/* CONFIG_PLINTH_NONE -- exclude the ANSI plinth
 *
 * Some MPS deployment environments want to avoid dependencies on the
 * standard C library.  In this case, the plinth, defined in mpslib.h must
 * be supplied when linking.
 *
 * For example, Open Dylan on Windows does not link the C library, but
 * supplies its own plinth directly using Windows and Dylan interfaces.
 *
 * CONFIG_PLINTH_NONE tells mps.c to exclude the ANSI plinth and removes
 * all standard C library dependencies.  e.g.
 *
 *     cc -O2 -c -DCONFIG_PLINTH_NONE mps.c
 */

#if defined(CONFIG_PLINTH_NONE)
#define PLINTH_NONE
#endif


#define MPS_VARIETY_STRING \
  MPS_ASSERT_STRING "." MPS_LOG_STRING "." MPS_STATS_STRING


/* Platform Configuration */

#include "mpstd.h"

/* Suppress Visual C warnings at warning level 4, */
/* see mail.richard.1997-09-25.13-26. */
/* Essentially the same settings are done in testlib.h. */

#ifdef MPS_BUILD_MV

/* "unreferenced inline function has been removed" (windows.h) */
#pragma warning(disable: 4514)

/* "constant conditional" (MPS_END) */
#pragma warning(disable: 4127)

/* "unreachable code" (ASSERT, if cond is constantly true). */
#pragma warning(disable: 4702)

/* "expression evaluates to a function which is missing an argument list" */
#pragma warning(disable: 4550)

/* "local variable is initialized but not referenced" */
#pragma warning(disable: 4189)

/* "not all control paths return a value" */
#pragma warning(disable: 4715)

/* MSVC 2.0 generates a warning when using NOCHECK or UNUSED */
#ifdef _MSC_VER
#if _MSC_VER < 1000
#pragma warning(disable: 4705)
#endif
#else /* _MSC_VER */
#error "Expected _MSC_VER to be defined for builder.mv"
#endif /* _MSC_VER */


/* Non-checking varieties give many spurious warnings because parameters
 * are suddenly unused, etc.  We aren't interested in these
 */

#if defined(AVER_AND_CHECK_NONE)

/* "unreferenced formal parameter" */
#pragma warning(disable: 4100)

/* "unreferenced local function has been removed" */
#pragma warning(disable: 4505)

#endif /* AVER_AND_CHECK_NONE */

#endif /* MPS_BUILD_MV */


/* EPVMDefaultSubsequentSegSIZE is a default for the alignment of
 * subsequent segments (non-initial at each save level) in EPVM.  See
 * design.mps.poolepvm.arch.segment.size.
 */

#define EPVMDefaultSubsequentSegSIZE ((Size)64 * 1024)


/* CBS Configuration -- see <code/cbs.c> */

#define CBS_EXTEND_BY_DEFAULT ((Size)4096)


/* Format defaults: see <code/format.c> */

#define FMT_ALIGN_DEFAULT ((Align)MPS_PF_ALIGN)
#define FMT_HEADER_SIZE_DEFAULT ((Size)0)
#define FMT_SCAN_DEFAULT (&FormatNoScan)
#define FMT_SKIP_DEFAULT (&FormatNoSkip)
#define FMT_FWD_DEFAULT (&FormatNoMove)
#define FMT_ISFWD_DEFAULT (&FormatNoIsMoved)
#define FMT_PAD_DEFAULT (&FormatNoPad)
#define FMT_CLASS_DEFAULT (&FormatDefaultClass)


/* Pool MV Configuration -- see <code/poolmv.c> */

#define MV_EXTEND_BY_DEFAULT  ((Size)65536)
#define MV_AVG_SIZE_DEFAULT   ((Size)32)
#define MV_MAX_SIZE_DEFAULT   ((Size)65536)


/* Pool MFS Configuration -- see <code/poolmfs.c> */

#define MFS_EXTEND_BY_DEFAULT ((Size)65536)


/* Pool MVFF Configuration -- see <code/poolmvff.c> */

#define MVFF_EXTEND_BY_DEFAULT   ((Size)65536)
#define MVFF_AVG_SIZE_DEFAULT    ((Size)32)
#define MVFF_ALIGN_DEFAULT       MPS_PF_ALIGN
#define MVFF_SLOT_HIGH_DEFAULT   FALSE
#define MVFF_ARENA_HIGH_DEFAULT  FALSE
#define MVFF_FIRST_FIT_DEFAULT   TRUE


/* Pool MVT Configuration -- see <code/poolmv2.c> */
/* FIXME: These numbers were lifted from mv2test and need thought. */

#define MVT_ALIGN_DEFAULT         MPS_PF_ALIGN
#define MVT_MIN_SIZE_DEFAULT      MPS_PF_ALIGN
#define MVT_MEAN_SIZE_DEFAULT     32
#define MVT_MAX_SIZE_DEFAULT      8192
#define MVT_RESERVE_DEPTH_DEFAULT 1024
#define MVT_FRAG_LIMIT_DEFAULT    30


/* Arena Configuration -- see <code/arena.c>
 *
 * .client.seg-size: ARENA_CLIENT_PAGE_SIZE is the size in bytes of a
 * "page" (i.e., segment granule) in the client arena.  It's set at 8192
 * with no particular justification.
 */

#define ArenaPollALLOCTIME (65536.0)

#define ARENA_ZONESHIFT         ((Shift)20)

#define ARENA_CLIENT_PAGE_SIZE          ((Size)8192)

#define ArenaDefaultZONESET (ZoneSetUNIV << (MPS_WORD_WIDTH / 2))
/* @@@@ knows the implementation of ZoneSets */

/* .segpref.default: For EPcore, non-DL segments should be placed high */
/* to reduce fragmentation of DL pools (see request.epcore.170193_). */
/* .. _request.epcore.170193: https://info.ravenbrook.com/project/mps/import/2001-11-05/mmprevol/request/epcore/170193 */
#define SegPrefDEFAULT { \
  SegPrefSig,          /* sig */ \
  TRUE,                /* high */ \
  ArenaDefaultZONESET, /* zoneSet */ \
  FALSE,               /* isCollected */ \
  FALSE,               /* isGen */ \
  (Serial)0,           /* gen */ \
}

#define LDHistoryLENGTH ((Size)4)

/* Value of MPS_KEY_EXTEND_BY for the arena control pool.
   Deliberately smaller than the default, because we don't expect the control
   pool to be very heavily used. */
#define CONTROL_EXTEND_BY 4096


/* Stack configuration */

/* Currently StackProbe has a useful implementation only on
 * Intel platforms and only when using Microsoft build tools (builder.mv)
 */
#if defined(MPS_ARCH_I3) && defined(MPS_BUILD_MV)
#define StackProbeDEPTH ((Size)500)
#else
#define StackProbeDEPTH ((Size)0)
#endif /* MPS_ARCH_I3 */


/* Shield Configuration -- see <code/shield.c> */

#define ShieldCacheSIZE ((size_t)16)
#define ShieldDepthWIDTH (4)


/* VM Configuration -- see <code/vm*.c> */

#define VMANPageALIGNMENT ((Align)4096)
#define VMJunkBYTE ((unsigned char)0xA9)
#define VMParamSize (sizeof(Word))


/* .feature.li: Linux feature specification
 *
 * The MPS needs the following symbols which are not defined if the
 * -ansi option is given to GCC:
 *
 * Source      Symbols                   Header        Feature
 * =========== ========================= ============= ====================
 * lockli.c    pthread_mutexattr_settype <pthread.h>   _XOPEN_SOURCE >= 500
 * prmci3li.c  REG_EAX etc.              <ucontext.h>  _GNU_SOURCE
 * prmci6li.c  REG_RAX etc.              <ucontext.h>  _GNU_SOURCE
 * prmcix.h    stack_t, siginfo_t        <signal.h>    _XOPEN_SOURCE
 * pthrdext.c  sigaction etc.            <signal.h>    _XOPEN_SOURCE
 * vmix.c      MAP_ANON                  <sys/mman.h>  _GNU_SOURCE
 *
 * It is not possible to localize these feature specifications around
 * the individual headers: all headers share a common set of features
 * (via <feature.h>) and so all sources in the same compilation unit
 * must turn on the same set of features.
 *
 * See "Feature Test Macros" in the Glibc Manual:
 * <http://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html>
 */

#if defined(MPS_OS_LI)

#define _XOPEN_SOURCE 500

#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif

#endif


/* .feature.xc: OS X feature specification
 *
 * The MPS needs the following symbols which are not defined by default
 *
 * Source      Symbols                   Header        Feature
 * =========== ========================= ============= ====================
 * prmci3li.c  __eax etc.                <ucontext.h>  _XOPEN_SOURCE
 * prmci6li.c  __rax etc.                <ucontext.h>  _XOPEN_SOURCE
 *
 * It is not possible to localize these feature specifications around
 * the individual headers: all headers share a common set of features
 * (via <sys/cdefs.h>) and so all sources in the same compilation unit
 * must turn on the same set of features.
 */

#if defined(MPS_OS_XC)

#if !defined(_XOPEN_SOURCE)
#define _XOPEN_SOURCE
#endif

#endif


/* Protection Configuration see <code/prot*.c>

   For each architecture/OS that uses protix.c or protsgix.c, we need to
   define what signal number to use, and what si_code value to check.
*/

#if defined(MPS_OS_FR)
#define PROT_SIGNAL (SIGSEGV)
#endif

#if defined(MPS_OS_FR)
#define PROT_SIGINFO_GOOD(info) ((info)->si_code == SEGV_ACCERR)
#endif


/* Almost all of protxc.c etc. are architecture-independent, but unfortunately
   the Mach headers don't provide architecture neutral symbols for simple
   things like thread states.  These definitions fix that. */

#if defined(MPS_OS_XC)
#if defined(MPS_ARCH_I6)

#define THREAD_STATE_COUNT x86_THREAD_STATE64_COUNT
#define THREAD_STATE_FLAVOR x86_THREAD_STATE64
#define THREAD_STATE_S x86_thread_state64_t

#elif defined(MPS_ARCH_I3)

#define THREAD_STATE_COUNT x86_THREAD_STATE32_COUNT
#define THREAD_STATE_FLAVOR x86_THREAD_STATE32
#define THREAD_STATE_S x86_thread_state32_t

#else

#error "Unknown OS X architecture"

#endif
#endif




/* Tracer Configuration -- see <code/trace.c> */

#define TraceLIMIT ((size_t)1)
/* I count 4 function calls to scan, 10 to copy. */
#define TraceCopyScanRATIO (1.5)

/* Chosen so that the RememberedSummaryBlockStruct packs nicely into
   pages */
#define RememberedSummaryBLOCK 15


/* Events
 *
 * EventBufferSIZE is the number of words in the global event buffer.
 */

#define EventBufferSIZE ((size_t)4096)
#define EventStringLengthMAX ((size_t)255) /* Not including NUL */


/* Assert Buffer */

#define ASSERT_BUFFER_SIZE      ((Size)512)


/* memory operator configuration
 *
 * We need efficient operators similar to memcpy, memset, and memcmp.
 * In general, we cannot use the C library mem functions directly as
 * that would not be freestanding.  However, on some platforms we can do
 * this, because they are inlined by the compiler and so do not actually
 * create a dependence on an external library.
 */

#if defined(MPS_PF_W3I3MV)
/* MSVC on Intel inlines mem* when optimizing */
#define mps_lib_memset(p, c, l) memset(p, c, l)
#define mps_lib_memcpy(p, q, s) memcpy(p, q, s)
#define mps_lib_memcmp(p, q, s) memcmp(p, q, s)
/* get prototypes for ANSI mem* */
#include <string.h>
#endif


/* Product Configuration
 *
 * Deprecated, see design/config/#req.prod>.  This now only contains the
 * configuration used by the former "MPS" product, which is now the only
 * product.
 */

#define MPS_PROD_STRING         "mps"
#define MPS_PROD_MPS
#define ARENA_INIT_SPARE_COMMIT_LIMIT   ((Size)10uL*1024uL*1024uL)
#define THREAD_MULTI
#define PROTECTION
#define PROD_CHECKLEVEL_INITIAL CheckLevelSHALLOW


/* Pool Class AMC configuration */

/* AMC treats segments of this many pages (or more) as "Large" */
#define AMCLargeSegPAGES ((Count)8)


/* Pool Class AWL configuration -- see poolawl.c for usage */

#define AWL_HAVE_SEG_SA_LIMIT   TRUE
#define AWL_SEG_SA_LIMIT        200     /* TODO: Improve guesswork with measurements */
#define AWL_HAVE_TOTAL_SA_LIMIT FALSE
#define AWL_TOTAL_SA_LIMIT      0


#endif /* config_h */


/* C. COPYRIGHT AND LICENSE
 *
 * Copyright (C) 2001-2013 Ravenbrook Limited <http://www.ravenbrook.com/>.
 * All rights reserved.  This is an open source license.  Contact
 * Ravenbrook for commercial licensing options.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 * 1. Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * 3. Redistributions in any form must be accompanied by information on how
 * to obtain complete source code for this software and any accompanying
 * software that uses this software.  The source code must either be
 * included in the distribution or be available for no more than the cost
 * of distribution plus a nominal fee, and must be freely redistributable
 * under reasonable conditions.  For an executable file, complete source
 * code means the source code for all modules it contains. It does not
 * include source code for modules or files that typically accompany the
 * major components of the operating system on which the executable file
 * runs.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
 * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
 * PURPOSE, OR NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDERS AND CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
